File: 10-01-iterator-graph/adts/linked_list.cpp
#pragma once

#include "./list.cpp"
#include <assert.h>

template <class T>
class linked_list : public list<T>, public iterable<T>
{
private:
  struct linked_list_node
  {
    T data;
    linked_list_node *next = nullptr;

    linked_list_node(T data)
    {
      this->data = data;
    }
  };

  class linked_list_iterator : public iterator<T>
  {
    linked_list_node *curr;

  public:
    linked_list_iterator(linked_list_node *start)
    {
      this->curr = start;
    }

    virtual bool hasNext() override
    {
      return this->curr != nullptr;
    }

    virtual T next() override
    {
      assert(this->curr != nullptr);
      T ret = this->curr->data;
      this->curr = this->curr->next;

      return ret;
    }
  };

  linked_list_node *head = nullptr;
  int length = 0;

public:
  linked_list()
  {
  }

  virtual iterator<T> *iterator() override
  {
    return new linked_list_iterator(this->head);
  }

  virtual void insert(T data) override
  {
    linked_list_node *newNode = new linked_list_node(data);
    if (this->head == nullptr)
    {
      head = newNode;
      this->length++;
      return;
    }
    // else head != nullptr

    linked_list_node *aux = this->head;
    while (aux->next != nullptr)
    {
      aux = aux->next;
    }

    aux->next = newNode;
    this->length++;
  }

  virtual int getPos(T data) override {}

  virtual T get(int pos) override
  {
    assert(pos < this->length);

    linked_list_node *aux = this->head;
    for (int i = 0; i < pos; i++)
    {
      aux = aux->next;
    }

    return aux->data;
  }

  virtual void remove(T data) override {}

  virtual void removeAt(int pos) override {}

  virtual int size() override
  {
    return this->length;
  }

  virtual bool isEmpty() override
  {
    return this->length == 0;
  }

  virtual bool contains(T data) override
  {
    linked_list_node *aux = this->head;

    while (aux != nullptr)
    {
      if (aux->data == data)
      {
        return true;
      }
      aux = aux->next;
    }

    return false;
  }
};

File: 10-01-iterator-graph/adts/LinkedList.java
package adts;

import java.util.Iterator;

public class LinkedList<T> implements List<T> {
  class LinkedListNode {
    T data;
    LinkedListNode next = null;

    LinkedListNode(T data) {
      this.data = data;
    }
  }

  class LinkedListIterator implements Iterator<T> {
    LinkedListNode curr;

    LinkedListIterator(LinkedListNode start) {
      this.curr = start;
    }

    @Override
    public boolean hasNext() {
      return this.curr != null;
    }

    @Override
    public T next() {
      T ret = this.curr.data;
      this.curr = this.curr.next;

      return ret;
    }
  };

  private LinkedListNode head = null;
  private int length = 0;

  @Override
  public void insert(T data) {
    LinkedListNode newNode = new LinkedListNode(data);
    if (this.head == null) {
      head = newNode;
      this.length++;
      return;
    }
    // else head != null

    LinkedListNode aux = this.head;
    while (aux.next != null) {
      aux = aux.next;
    }

    aux.next = newNode;
    this.length++;
  }

  @Override
  public int getPos(T data) {
    // TODO Auto-generated method stub
    throw new UnsupportedOperationException("Unimplemented method 'getPos'");
  }

  @Override
  public T get(int pos) {
    if (pos >= this.length) {
      throw new RuntimeException("out of range");
    }

    LinkedListNode aux = this.head;
    for (int i = 0; i < pos; i++) {
      aux = aux.next;
    }

    return aux.data;
  }

  @Override
  public void remove(T data) {
    // TODO Auto-generated method stub
    throw new UnsupportedOperationException("Unimplemented method 'remove'");
  }

  @Override
  public void removeAt(int pos) {
    // TODO Auto-generated method stub
    throw new UnsupportedOperationException("Unimplemented method 'removeAt'");
  }

  @Override
  public int size() {
    return this.length;
  }

  @Override
  public boolean isEmpty() {
    return this.length == 0;
  }

  @Override
  public boolean contains(T data) {
    LinkedListNode aux = this.head;

    while (aux != null) {
      if (aux.data.equals(data)) {
        return true;
      }
      aux = aux.next;
    }

    return false;
  }

  @Override
  public Iterator<T> iterator() {
    return new LinkedListIterator(this.head);
  }

}
